### ISSUE TITLE
Develop coding style guide

### ISSUE THREAD

**Issue by jschwab** [MESAHub/mesa#76]
It would be useful to have a coding style guide.  This could cover information about formatting as well as style (do loops vs arrays).

**Comment by matthiasfabry** [comment]
I'd like to bump this by saying lots of `.f90` files aren't formatted free-form style, but rather fixed-form style.
I'd argue free-from dramatically improves readability, e.g. look at `star/public/star_def.f90`:
```
      module star_def
      
      use star_data_def

      
      implicit none      
      
      
      end module star_def
```
vs.
```
module star_def

   use star_data_def

   implicit none

end module star_def
```
I might be overly fanatic about this, but I feel the code base could benefit from this. I can try a bulk reformat if this is something we'd like to see happen.

**Comment by rhdtownsend** [comment]
I wholeheartedly approve of such a change; much of the current source code is (visually) an abomination.

**Comment by rjfarmer** [comment]
@matthiasfabry if you are going to insist on whole codebase changes please only change the Fortran files. I've seen your [free-format-formatting](https://github.com/MESAHub/mesa/tree/free-form-formatting) branch and you are changing every file including those that aren't Fortran files. It's going to be difficult to merge if you insist on changing every file at once. Instead, go for one module at a time so there is even a slight chance we can review changes before merging.

Please also don't change the code in the stella folder. No one here can fix things if you break something in it.

**Comment by matthiasfabry** [comment]
yes, that makes sense from a review standpoint. I can restart and make a folder of branches.

Also good to know `stella` is off limits

**Comment by matthiasfabry** [comment]
I've created formatting/astero, /auto_diff, /binary and /const for you to review. I've tried to be as careful as possible for the formatter to only touch `.f90` files.

**Comment by evbauer** [comment]
The fortran code in `auto_diff` is actually generated by python scripts from the `python` directory in that module, so we should generally avoid direct edits to the `auto_diff/private` files. If you need to make edits to `auto_diff`, you should edit python files and then regenerate the fortran as necessary.

I guess this isn't particularly well documented either, so maybe we should try to add some documentation somewhere to help developers navigate editing `auto_diff` if necessary. I'll admit I'm not very clear on that at the moment.

**Comment by earlbellinger** [comment]
Here's some documentation on auto_diff:
https://adamjermyn.com/posts/auto_diff_mesa/

Best wishes,
Earl

**Comment by evbauer** [comment]
Nice! I didn't realize there was so much nice documentation out there. We probably shouldn't rely too heavily on @adamjermyn's blog for hosting important documentation though, so I wonder if there's an easy way to pull that page into the MESA website as part of our `auto_diff` documentation page [here](https://docs.mesastar.org/en/latest/auto_diff/overview.html).

@adamjermyn do you happen to have source that would be compatible the the ReST format of MESA docs? Or should we just start by copying over [the html](https://github.com/adamjermyn/adamjermyn.github.io/blob/master/posts/auto_diff_mesa/index.html)?

**Comment by adamjermyn** [comment]
The blog post is auto-generated with Hugo from the attached markdown file.
Should be easy enough to get it into ReST format!

-Adam

**Comment by evbauer** [comment]
Thanks! A markdown file should be super easy to port over. I don't see the attachment though...

**Comment by fxt44** [comment]
awesome! thank you evan.
